File: src/main/java/ch/obermuhlner/astro/gradient/filter/HorizontalMedianFilter.java
package ch.obermuhlner.astro.gradient.filter;

import ch.obermuhlner.astro.image.DoubleImage;
import ch.obermuhlner.astro.image.color.ColorModel;

import java.util.Arrays;
import java.util.Comparator;

public class HorizontalMedianFilter implements Filter {

  private final int radius;
  private final ColorModel model;

  public HorizontalMedianFilter(int radius, ColorModel model) {
    this.radius = radius;
    this.model = model;
  }

  @Override
  public DoubleImage filter(DoubleImage source, DoubleImage target, int width, int height) {
    double[][] data = new double[width][3];

    for (int y = 0; y < height; y++) {
      for (int x = 0; x < width; x++) {
        int n = 0;
        for (int dx = -radius; dx <= radius; dx++) {
          int xx = x + dx;
          if (source.isReallyInside(xx, y)) {
            source.getPixel(xx, y, model, data[n++]);
          }
        }
        Arrays.sort(data, 0, n, Comparator.<double[]>
            comparingDouble(c -> c[2])
            .thenComparing(c -> c[1])
            .thenComparing(c -> c[0]));

        target.setPixel(x, y, model, data[n/2]);
      }
    }
    return target;
  }

  @Override
  public String toString() {
    return "HorizontalMedianFilter(radius=" + radius + ")";
  }
}

File: src/main/java/ch/obermuhlner/astro/gradient/filter/HorizontalMedianFilter.kt
package ch.obermuhlner.astro.gradient.filter

import ch.obermuhlner.astro.image.DoubleImage
import ch.obermuhlner.astro.image.color.ColorModel
import java.util.*
import java.util.function.Function
import java.util.function.ToDoubleFunction

class HorizontalMedianFilter constructor(private val radius: Int, private val model: ColorModel) : Filter {
    override fun filter(source: DoubleImage, target: DoubleImage, width: Int, height: Int): DoubleImage {
        val data: Array<DoubleArray> = Array(width) { DoubleArray(3) }
        for (y in 0 until height) {
            for (x in 0 until width) {
                var n = 0
                for (dx in -radius..radius) {
                    val xx: Int = x + dx
                    if (source.isReallyInside(xx, y)) {
                        source.getPixel(xx, y, model, data.get(n++))
                    }
                }
                Arrays.sort(data, 0, n,
                        Comparator.comparingDouble { c: DoubleArray -> c[2] }
                        .thenComparing { c: DoubleArray -> c[1] }
                        .thenComparing { c: DoubleArray -> c[0] })
                target.setPixel(x, y, model, data[n / 2])
            }
        }
        return target
    }

    override fun toString(): String {
        return "HorizontalMedianFilter(radius=$radius)"
    }
}